sbncode/TPCReco/CNNHitClassification/CNNID_module.cc

```diff
@@ -0,0 +1,318 @@
+    std::vector<art::Ptr<recob::Hit>> hitPtrList;
+    art::fill_ptr_vector(hitPtrList, hitListHandle);
+
+    auto pfpListHandle = evt.getValidHandle<std::vector<recob::PFParticle>>(fPFParticleModuleLabel);
+		std::vector<art::Ptr<recob::PFParticle>> pfpList;
+    art::fill_ptr_vector(pfpList, pfpListHandle);
+
+    auto metaHandle = evt.getValidHandle<std::vector<larpandoraobj::PFParticleMetadata>>(fPFParticleModuleLabel);
+    std::vector<art::Ptr<larpandoraobj::PFParticleMetadata>> metaPtrList;
+    art::fill_ptr_vector(metaPtrList, metaHandle);
+
+    auto cluListHandle = evt.getValidHandle<std::vector<recob::Cluster>>(fClusterModuleLabel);
+    std::vector<art::Ptr<recob::Cluster>> cluPtrList;
+    art::fill_ptr_vector(cluPtrList, cluListHandle);
+
+    art::FindManyP<larpandoraobj::PFParticleMetadata> metaFMpfp(pfpListHandle, evt, fPFParticleModuleLabel);
+
+    art::FindManyP<recob::Cluster> cluFMpfp(pfpListHandle, evt, fClusterModuleLabel);
+
+    art::FindManyP<recob::Hit> hitFMclu(cluListHandle, evt, fClusterModuleLabel);
+    art::FindManyP<recob::Hit> hitFMpfp(pfpListHandle, evt, fPFParticleModuleLabel);
+
+    auto hitID = fMVAWriter.initOutputs<recob::Hit>(
+      fHitModuleLabel, hitPtrList.size(), fPointIdAlg.outputLabels());
+
+    auto pfpScores = std::make_unique<std::vector<PFPCNNScore>>();
+    auto pfpAssns = std::make_unique<art::Assns<recob::PFParticle, PFPCNNScore>>();
+
+    // loop over PFPs
+    for (const art::Ptr<recob::PFParticle> &thispfp : pfpList){
+      std::vector<float> PFPTrackScore; 
+      std::vector<float> PFPShowerScore; 
+      std::vector<float> PFPNoiseScore;
+      // std::vector<float> PFPMichelScore;
+      int nClusters = 0; 
+
+      // skip clear cosmic PFPs
+      std::vector<art::Ptr<larpandoraobj::PFParticleMetadata>> metas = metaFMpfp.at(thispfp.key());
+      const art::Ptr<larpandoraobj::PFParticleMetadata> meta = metas[0];
+      auto const &properties = meta->GetPropertiesMap();
+      bool is_clear_cosmic = (properties.count("IsClearCosmic")); 
+      if (is_clear_cosmic) continue;
+
+      // loop over clusters
+      auto const &clusters = cluFMpfp.at(thispfp->Self());
+      for (auto const &cluster : clusters){
+        float cluTrackScore = 0.;
+        float cluShowerScore = 0.;
+        float cluNoiseScore = 0.;
+        // float cluMichelScore = 0.;
+
+        auto &cluHitPtrList = hitFMclu.at(cluster.key());
+        int nHitsClu = cluHitPtrList.size();
+        if (nHitsClu < 5) continue; // at least 5 hits in cluster
+        nClusters++;
+
+        cryo = cluster->Plane().Cryostat;
+        view = cluster->Plane().Plane;
+        tpc = cluster->Plane().TPC;
+        fPointIdAlg.setWireDriftData(clockData, detProp, *wireHandle, view, tpc, cryo);
+
+        // hits to apply CNN to (sparsely if nHitsClu > fSparseLengthCut)
+        std::vector<art::Ptr<recob::Hit>> cluHitPtrListApply; 
+        if (nHitsClu > fSparseLengthCut) {
```

Comment:
It looks like clusters below a certain NHits threshold are ignored.  Is this deliberate?